### ISSUE TITLE
Turn `twisted.protocols.ftp.SenderProtocol` into a regular `IConsumer`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4580]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4580|
|Type|task|
|Created|2010-07-11 18:28:44Z|
|Branch|https://github.com/twisted/twisted/tree/convert-SenderProtocol-4580-2|

`SenderProtocol` implements `IFinishableConsumer` which is going to be deprecated ([#2661](https://github.com/twisted/twisted/issues/2661)).  We should change this to be a plain `IConsumer`.

<details><summary>Searchable metadata</summary>

```
trac-id__4580 4580
type__task task
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_convert_SenderProtocol_4580_2 branches/convert-SenderProtocol-4580-2
branch_author__cyli__exarkun cyli, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1278872924000000 1278872924000000
changetime__1300052325000000 1300052325000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

It seems there are several possible approaches here:

  1. Move `IFinishableConsumer` into ftp.py and just treat it as an ftp-specific interface.  Let `SenderProtocol` continue to implement it.
  1. Delete the declaration in `SenderProtocol` that the class implements `IFinishableConsumer`.  Change any documentation which talks about an object that provides `IFinishableConsumer` to instead talk about an object that provides `IFinishableConsumer` and also has a `finish` method.
  1. Delete the declaration in `SenderProtocol` that the class implements `IFinishableConsumer`.  Where an `IFinishableConsumer` was being returned (typically as part of a tuple), return an extra finish callable as a separate object.
  1. Invert the interfaces to accept a producer instead of returning a consumer.  Define some interface to hook arbitrary producers up to consumers.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

(In [30433]) Branching to 'convert-SenderProtocol-4580'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli commented|
|-|-|

(In [30434]) Changed SenderProtocol to not implement interfaces.IFinishableConsumer

refs [#4580](https://github.com/twisted/twisted/issues/4580)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli removed owner|
|-|-|

I went for [#11338](https://github.com/twisted/twisted/issues/11338) because it was easiest.

Build Results:  http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/convert-SenderProtocol-4580

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra set owner to @cyli|
|-|-|

`IFinishableConsumer` needs an `@since` marker.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

(In [30455]) Added @since marker to IFinishableConsumer

refs [#4580](https://github.com/twisted/twisted/issues/4580)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra commented|
|-|-|

Replying to [cyli](#note_6):
> (In [30455]) Added @since marker to IFinishableConsumer
> 
> refs [#4580](https://github.com/twisted/twisted/issues/4580)

Sorry to be anal about this but I think the syntax is `@since: 11.0` :)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

(In [30459]) Changed the @since marker text

refs [#4580](https://github.com/twisted/twisted/issues/4580)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set owner to @cyli|
|-|-|

Almost there: '@since: 11.0', not '@since 11.0'

Please merge once fixed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [30984]) Branching to 'convert-SenderProtocol-4580-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [30986]) Fix epytext markup

refs [#4580](https://github.com/twisted/twisted/issues/4580)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [30987]) Merge convert-SenderProtocol-4580-2

Author: cyli
Reviewer: therve
Fixes: [#4580](https://github.com/twisted/twisted/issues/4580)

Add `IFinishableConsumer` to the ftp module, the only user of it, so that it can
be removed from the more general `twisted.internet.interfaces` module.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `reopened`|
|-|-|

(In [30991]) Revert r30987 - api doc regression caused by mis-merge

Reopens [#4580](https://github.com/twisted/twisted/issues/4580)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set status to `closed`|
|-|-|

(In [30992]) Merge convert-SenderProtocol-4580-2

Author: cyli
Reviewer: therve
Fixes: [#4580](https://github.com/twisted/twisted/issues/4580)

Add `IFinishableConsumer` to the ftp module, the only user of it, so that it can
be removed from the more general `twisted.internet.interfaces` module.

This commit should actually properly merge the branch, instead of a random
older revision of it.
